GeckoWebExecutor implementation

keianhzo opened the pull request:
@andrenatal This implementation adds support for a GeckoWebExecutor based Http connection so projects already using Gecko can simplify the Http stack. The problem is that I get an invalid answer from the server but I cannot debug it so I'm not sure what's happening or why the request is not valid.

arybin commented on mozillaspeechlibrary/src/main/java/com/mozilla/speechlibrary/networking/CustomNetworking.java:
```diff
@@ -0,0 +1,93 @@
+            );
+
+        } catch(Exception e)  {
+            e.printStackTrace();
```
consider reporting this exception to the `mSpeechService`

andrenatal replied:
I second this

keianhzo replied:
Fixed

arybin commented on mozillaspeechlibrary/src/main/java/com/mozilla/speechlibrary/networking/NetworkSettings.java:
```diff
@@ -0,0 +1,9 @@
+package com.mozilla.speechlibrary.networking;
+
+public class NetworkSettings {
```
qq: why not introduce Kotlin in the codebase for handling data classes like this?

andrenatal commented on mozillaspeechlibrary/src/main/java/com/mozilla/speechlibrary/networking/GeckoNetworking.java:
```diff
@@ -0,0 +1,94 @@
+            input.put(baos.toByteArray());
+            WebRequest request = new WebRequest.Builder(STT_ENDPOINT)
+                    .body(input)
+                    .addHeader("Accept-Language-STT", mNetworkSettings.mLanguage)
```
After adding `.method("POST")` here, and debugging locally along the server, I started seeing the proper json containing the transcription being returned

andrenatal commented on mozillaspeechlibrary/src/main/java/com/mozilla/speechlibrary/networking/GeckoNetworking.java:
```diff
@@ -0,0 +1,94 @@
+            input.put(baos.toByteArray());
+            WebRequest request = new WebRequest.Builder(STT_ENDPOINT)
+                    .body(input)
+                    .addHeader("Accept-Language-STT", mNetworkSettings.mLanguage)
```
```suggestion
                    .method("POST")
                    .addHeader("Accept-Language-STT", mNetworkSettings.mLanguage)
```

keianhzo replied:
Added, I had it in the past, I think this PR didn't have it because I was just trying anything to make it work.

andrenatal commented on mozillaspeechlibrary/src/main/java/com/mozilla/speechlibrary/networking/GeckoNetworking.java:
```diff
@@ -0,0 +1,94 @@
+                String body;
+                if (webResponse != null) {
+                    if (webResponse.body != null) {
+                        int size = webResponse.body.available();
```
But even after applying the changes aforementioned and having confirmed that the json is returned from the server to the client using Wireshark, this always return 0. So seems to me that the executor is not reading the result properly.

bluemarvin replied:
I from https://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html#available()

> Note that while some implementations of InputStream will return the total number of bytes in the stream, many will not. It is never correct to use the return value of this method to allocate a buffer intended to hold all data in this stream. 

Just call `read()`

keianhzo replied:
Moving it to read worked fine. Fixed.

keianhzo commented:
@andrenatal I've fixed the issue, made some changes and updated to AndroidX. I think this is ready for review, I've tested it in FxR and it works fine. You can test it in FxR using the `v8/mozillaspeech_gbe` although I'm still woring on it to remove all the related http stack dependencies.

keianhzo commented:
@andrenatal You should be able to test this in FxR now: https://github.com/MozillaReality/FirefoxReality/pull/2886

arybin commented on mozillaspeechlibrary/src/main/java/com/mozilla/speechlibrary/networking/CustomNetworking.java:
```diff
@@ -0,0 +1,92 @@
+                            // called when response HTTP status is "200 OK"
+                            String json = new String(response);
+                            try {
+                                JSONObject reader = new JSONObject(json);
```
it seems that this code is duplicated here and in  `GeckoNetworking.java` consider extracting out into a helper class for parsing JSON into `STTResult`.

arybin commented on mozillaspeechlibrary/src/main/java/com/mozilla/speechlibrary/networking/GeckoNetworking.java:
```diff
@@ -0,0 +1,100 @@
+                    if (webResponse.body != null) {
+                        ByteArrayOutputStream buffer = new ByteArrayOutputStream();
+                        int nRead;
+                        byte[] data = new byte[16384];
```
magic number?
